File: tests/typescript/unit/sdk/context.test.ts
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// Unit tests for SDK context propagation: bind/current, capture, overrides, and envelope round-trips.

import { describe, it, expect } from 'vitest'
import {
  bind,
  current,
  captureContext,
  withOverrides,
  toEnvelope,
  fromEnvelope,
  describeAuthority,
  type CaracalContext,
} from '../../../../packages/sdk/ts/src/context.js'

function ctx(overrides: Partial<CaracalContext> = {}): CaracalContext {
  return {
    subjectToken: 'tok',
    zoneId: 'zone-1',
    applicationId: 'app-1',
    sessionId: 'agent-1',
    subjectAuthorityRecordId: 'sess-1',
    traceId: 'trace-1',
    hop: 0,
    ...overrides,
  }
}

describe('bind and current', () => {
  it('exposes the bound context inside the callback and clears it outside', async () => {
    expect(current()).toBeUndefined()
    const seen = await bind(ctx(), async () => current())
    expect(seen?.sessionId).toBe('agent-1')
    expect(current()).toBeUndefined()
  })

  it('isolates concurrent bind scopes from each other', async () => {
    const seen = await Promise.all(
      ['a', 'b', 'c'].map((id) =>
        bind(ctx({ sessionId: `agent-${id}` }), async () => {
          await new Promise((resolve) => setTimeout(resolve, 1))
          return current()?.sessionId
        }),
      ),
    )
    expect(seen).toEqual(['agent-a', 'agent-b', 'agent-c'])
  })
})

describe('captureContext', () => {
  it('returns undefined with no active context', () => {
    expect(captureContext()).toBeUndefined()
  })

  it('returns a detached copy of the active context', async () => {
    await bind(ctx(), async () => {
      const snap = captureContext()
      expect(snap).toEqual(current())
      expect(snap).not.toBe(current())
    })
  })

  it('clones baggage so the snapshot cannot mutate the bound context', async () => {
    await bind(ctx({ baggage: { tenant: 'piedpiper' } }), async () => {
      const snap = captureContext()
      snap!.baggage!.tenant = 'hooli'
      expect(current()?.baggage?.tenant).toBe('piedpiper')
    })
  })
})

describe('withOverrides', () => {
  it('throws when no base context exists', () => {
    expect(() => withOverrides({ hop: 9 })).toThrow(/requires an existing/)
  })

  it('returns a merged copy without mutating the bound context', async () => {
    await bind(ctx(), async () => {
      const merged = withOverrides({ hop: 5, applicationId: 'app-2' })
      expect(merged.hop).toBe(5)
      expect(merged.applicationId).toBe('app-2')
      expect(merged.zoneId).toBe('zone-1')
      expect(current()?.hop).toBe(0)
      expect(current()?.applicationId).toBe('app-1')
    })
  })

  it('clones baggage so the copy cannot mutate the base', async () => {
    await bind(ctx({ baggage: { tenant: 'piedpiper' } }), async () => {
      const merged = withOverrides({ hop: 1 })
      merged.baggage!.tenant = 'hooli'
      expect(current()?.baggage?.tenant).toBe('piedpiper')
    })
  })
})

describe('envelope round-trip', () => {
  it('serializes and restores a context via the envelope', () => {
    const original = ctx({ delegationId: 'edge-1', parentDelegationId: 'edge-0', hop: 2 })
    const env = toEnvelope(original)
    const restored = fromEnvelope(env, { zoneId: 'zone-1', applicationId: 'app-1' })
    expect(restored).toMatchObject({
      subjectToken: 'tok',
      sessionId: 'agent-1',
      delegationId: 'edge-1',
      parentDelegationId: 'edge-0',
      hop: 2,
    })
  })

  it('rejects an envelope without a subject token', () => {
    expect(() => fromEnvelope({ hop: 0 } as never, { zoneId: 'z', applicationId: 'c' })).toThrow(/missing subject token/)
  })
})

describe('describeAuthority', () => {
  it('returns undefined without a context', () => {
    expect(describeAuthority(undefined)).toBeUndefined()
  })

  it('builds the full authority chain in order', () => {
    const summary = describeAuthority(
      ctx({
        delegationId: 'edge-1',
        parentDelegationId: 'edge-0',
        hop: 3,
      }),
    )
    expect(summary?.chain).toEqual(['subject:sess-1', 'session:agent-1', 'parent-delegation:edge-0', 'delegation:edge-1'])
    expect(summary).toMatchObject({ zoneId: 'zone-1', applicationId: 'app-1', hop: 3 })
  })

  it('omits chain segments for absent identifiers', () => {
    const summary = describeAuthority(ctx({ sessionId: undefined, subjectAuthorityRecordId: undefined }))
    expect(summary?.chain).toEqual([])
  })
})

File: packages/sdk/go/context_test.go
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// Unit tests for CaracalContext bind/current and envelope projection.

package sdk_test

import (
	"context"
	"runtime"
	"strings"
	"sync"
	"testing"

	sdk "github.com/garudex-labs/caracal/packages/sdk/go"
)

func TestBindCurrentRoundTrip(t *testing.T) {
	c := sdk.CaracalContext{
		SubjectToken:             "tok",
		ZoneID:                   "z1",
		ApplicationID:            "app1",
		SessionID:                "sess",
		DelegationID:             "edge",
		SubjectAuthorityRecordID: "sid",
		Hop:                      2,
	}
	ctx := sdk.Bind(context.Background(), c)
	got, ok := sdk.Current(ctx)
	if !ok {
		t.Fatal("Current must return true after Bind")
	}
	if got.SubjectToken != c.SubjectToken {
		t.Errorf("SubjectToken: %q vs %q", got.SubjectToken, c.SubjectToken)
	}
	if got.ZoneID != c.ZoneID {
		t.Errorf("ZoneID: %q vs %q", got.ZoneID, c.ZoneID)
	}
	if got.Hop != c.Hop {
		t.Errorf("Hop: %d vs %d", got.Hop, c.Hop)
	}
}

func TestCurrentOnFreshContext(t *testing.T) {
	_, ok := sdk.Current(context.Background())
	if ok {
		t.Error("Current must return false on a context with no Bind")
	}
}

func TestCaptureReturnsCurrentContext(t *testing.T) {
	want := sdk.CaracalContext{SubjectToken: "tok", ZoneID: "z1", ApplicationID: "app1"}
	got, ok := sdk.Capture(sdk.Bind(context.Background(), want))
	if !ok {
		t.Fatal("Capture must return true after Bind")
	}
	if got.SubjectToken != want.SubjectToken || got.ZoneID != want.ZoneID || got.ApplicationID != want.ApplicationID {
		t.Fatalf("Capture = %#v, want %#v", got, want)
	}
}

func TestBindDoesNotMutateParent(t *testing.T) {
	parent := context.Background()
	sdk.Bind(parent, sdk.CaracalContext{SubjectToken: "tok"})
	_, ok := sdk.Current(parent)
	if ok {
		t.Error("Bind must not modify the parent context")
	}
}

func TestBindClonesBaggage(t *testing.T) {
	baggage := map[string]string{"tenant": "piedpiper"}
	ctx := sdk.Bind(context.Background(), sdk.CaracalContext{SubjectToken: "tok", Baggage: baggage})
	baggage["tenant"] = "hooli"

	bound, _ := sdk.Current(ctx)
	if bound.Baggage["tenant"] != "piedpiper" {
		t.Errorf("bound baggage = %q, want isolation from caller mutation", bound.Baggage["tenant"])
	}

	bound.Baggage["tenant"] = "endframe"
	again, _ := sdk.Current(ctx)
	if again.Baggage["tenant"] != "piedpiper" {
		t.Errorf("stored baggage = %q, want isolation from reader mutation", again.Baggage["tenant"])
	}
}

func TestBindIsolatesConcurrentGoroutines(t *testing.T) {
	agents := []string{"agent-a", "agent-b", "agent-c"}
	results := make([]string, len(agents))
	var wg sync.WaitGroup
	for i, agent := range agents {
		wg.Add(1)
		go func() {
			defer wg.Done()
			ctx := sdk.Bind(context.Background(), sdk.CaracalContext{SubjectToken: "tok", SessionID: agent})
			runtime.Gosched()
			got, _ := sdk.Current(ctx)
			results[i] = got.SessionID
		}()
	}
	wg.Wait()
	for i, agent := range agents {
		if results[i] != agent {
			t.Errorf("goroutine %d saw %q, want %q", i, results[i], agent)
		}
	}
}

func TestFromEnvelopeFullFields(t *testing.T) {
	env := sdk.Envelope{
		SubjectToken:             "tok",
		SessionID:                "sess",
		DelegationID:             "edge",
		ParentDelegationID:       "parent",
		SubjectAuthorityRecordID: "sid",
		TraceID:                  "0123456789abcdef0123456789abcdef",
		Hop:                      4,
	}
	c, err := sdk.FromEnvelope(env, "zone1", "app1")
	if err != nil {
		t.Fatal(err)
	}
	if c.SubjectToken != "tok" {
		t.Errorf("SubjectToken: %q", c.SubjectToken)
	}
	if c.ZoneID != "zone1" {
		t.Errorf("ZoneID: %q", c.ZoneID)
	}
	if c.SessionID != "sess" {
		t.Errorf("SessionID: %q", c.SessionID)
	}
	if c.SubjectAuthorityRecordID != "sid" {
		t.Errorf("SessionID: %q", c.SubjectAuthorityRecordID)
	}
	if c.Hop != 4 {
		t.Errorf("Hop: %d", c.Hop)
	}
}

func TestFromEnvelopeMissingSubjectTokenErrors(t *testing.T) {
	_, err := sdk.FromEnvelope(sdk.Envelope{}, "z", "a")
	if err == nil {
		t.Fatal("expected error for missing subject token")
	}
}

func TestToEnvelopeRoundTrip(t *testing.T) {
	c := sdk.CaracalContext{
		SubjectToken:             "tok",
		SessionID:                "sess",
		DelegationID:             "edge",
		ParentDelegationID:       "parent",
		SubjectAuthorityRecordID: "sid",
		TraceID:                  "0123456789abcdef0123456789abcdef",
		Hop:                      2,
	}
	env := sdk.ToEnvelope(c)
	if env.SubjectToken != c.SubjectToken {
		t.Errorf("SubjectToken: %q vs %q", env.SubjectToken, c.SubjectToken)
	}
	if env.SessionID != c.SessionID {
		t.Errorf("SessionID: %q vs %q", env.SessionID, c.SessionID)
	}
	if env.TraceID != c.TraceID {
		t.Errorf("TraceID: %q vs %q", env.TraceID, c.TraceID)
	}
	if env.SubjectAuthorityRecordID != c.SubjectAuthorityRecordID {
		t.Errorf("SessionID: %q vs %q", env.SubjectAuthorityRecordID, c.SubjectAuthorityRecordID)
	}
	if env.Hop != c.Hop {
		t.Errorf("Hop: %d vs %d", env.Hop, c.Hop)
	}
}

func TestToEnvelopeFromEnvelopeRoundTrip(t *testing.T) {
	orig := sdk.CaracalContext{
		SubjectToken:             "tok",
		ZoneID:                   "z",
		ApplicationID:            "app",
		SessionID:                "sess",
		DelegationID:             "edge",
		SubjectAuthorityRecordID: "sid",
		Hop:                      1,
	}
	env := sdk.ToEnvelope(orig)
	restored, err := sdk.FromEnvelope(env, orig.ZoneID, orig.ApplicationID)
	if err != nil {
		t.Fatal(err)
	}
	if restored.SubjectToken != orig.SubjectToken {
		t.Errorf("SubjectToken mismatch")
	}
	if restored.SessionID != orig.SessionID {
		t.Errorf("SessionID mismatch")
	}
	if restored.SubjectAuthorityRecordID != orig.SubjectAuthorityRecordID {
		t.Errorf("SessionID mismatch")
	}
}

func TestDescribeAuthorityRedactsSubjectToken(t *testing.T) {
	ctx := sdk.Bind(context.Background(), sdk.CaracalContext{
		SubjectToken:             "tok",
		ZoneID:                   "z",
		ApplicationID:            "app",
		SubjectAuthorityRecordID: "sid",
		SessionID:                "agent",
		DelegationID:             "edge",
		Hop:                      2,
	})
	summary, ok := sdk.DescribeAuthority(ctx)
	if !ok {
		t.Fatal("DescribeAuthority must return a summary for a bound context")
	}
	if summary.ApplicationID != "app" || summary.SubjectAuthorityRecordID != "sid" || summary.SessionID != "agent" {
		t.Fatalf("unexpected summary: %#v", summary)
	}
	got := strings.Join(summary.Chain, ">")
	want := "subject:sid>session:agent>delegation:edge"
	if got != want {
		t.Fatalf("chain = %q, want %q", got, want)
	}
}

func TestDescribeAuthorityIncludesParentDelegationAndFreshContextFalse(t *testing.T) {
	if _, ok := sdk.DescribeAuthority(context.Background()); ok {
		t.Fatal("fresh context should not describe authority")
	}
	summary := sdk.DescribeContext(sdk.CaracalContext{
		SubjectAuthorityRecordID: "sid",
		SessionID:                "agent",
		ParentDelegationID:       "parent-edge",
		DelegationID:             "edge",
	})
	got := strings.Join(summary.Chain, ">")
	want := "subject:sid>session:agent>parent-delegation:parent-edge>delegation:edge"
	if got != want {
		t.Fatalf("chain = %q, want %q", got, want)
	}
}
